design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a dependency caching mechanism that allows the outputs of steps within a pipeline to be optimized (or even skipped) if a suitable cached version of the outputs of those steps already exists. Caching can be effective provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch.
+
+## Proposal
+
+This proposal contains a number of interconnected elements. We have attempted to layer them in this document from the most primitive building blocks up to simplified YAML syntax to make caching "just work" in the majority of cases. Our approach to building a caching mechanism for Azure Pipelines will be to get the fundamental building blocks right and prove we can increase build performance and then evolve the YAML syntax once we know exactly what that syntax will need to describe.
+
+### Restore and Save Cache Tasks
+
+The basic functions of Pipeline Cache will be implemented by two tasks. The ```RestoreCache``` task and the ```SaveCache``` task. Both tasks will specify a collection of files and environment variables which can be used as a key to lookup the cache and identify a hit or a miss. Each task will also take a collection of paths which will be fetched from or stored in the cache associated with that key.
```

Comment:
Will the cache be immutable or will it be possible to be updated for the same key? The reason why I am asking this is that for maven it would be needed to use all `pom.xml` within all the multi module project as a hash key, in case the cache is immutable. The main pom.xml is not enough as you might add a managed dependency imported via a bom and the cache will never get updated.

Is there some other possibility for caching maven artifacts?

Reply:
The cache will be immutable, however there might be another solution here. Where we specify the key we could potentially take a globbing syntax to make sure we find all your pom files?

```
- saveCache:
  key:
    **/pom.xml
  paths:
    foo
    bar
```

Would that work for your scenario?

Reply:
If we can glob on all files that could work. However, for maven that can lead to unneeded cache miss, for example when you add a dependency that is already in the cache, or if you add your own dependency from the reactor.

In any case this could also be analysed on your side depending on the usage. It is also really dependent on the size of the project

Reply:
If we can glob on all files that could work. However, for maven that can lead to unneeded cache miss, for example when you add a dependency that is already in the cache, or if you add your own dependency from the reactor.

In any case this could also be analysed on your side depending on the usage. It is also really dependent on the size of the project

Reply:
I think as we get into tighter integration with particular ecosystems we have some opportunities to adapt to their specific scenarios. Thanks for giving me something to think about...

A cache miss when any dependencies change is kind of by design. In those scenarios we expect the build to take longer to restore dependencies but in doing so the resulting cache is more "correct".

Reply:
The problem with the approach for Maven is that if you hash all `pom.xml` and you change something (e.g. plugin configuration) that does not affect or keeps the cache as is you will get a cache miss.

For example I have the following structure:

* parent
  * module1
  * module2

Module1 already depends on some library, this library is in the Maven cache (.m2) repository. After the creation of the cache I add the same library to `module2` then the cache will be invalidated, but for Maven nothing will change, it won't need to download the artifact again as it is already present in its cache (.m2) repository. 

In any case I find it really good the way you are interacting with the community and discussing new features in the open like this.

Reply:
Hey thanks :)

Do you have a feel for how often you might edit change a plug-in configuration vs. regular source changes in the Java ecosystem. Whilst a cache miss will result in longer builds its not like the output of the build will be incorrect, it'll just take a bit longer in those particular instances.

So if a cache miss occurs 10% of the time, but the other 90% is OK, the outcome might be "good enough"?